### PR TITLE
Fixes #35572 - Fails publish if repos have no content

### PR DESCRIPTION
**Pull request by parthaa**
#### What are the changes introduced in this pull request?
Customers who have upgraded from katello 3.18 and above in some instances seem to have repositories that do not have any candlepin content. This is possibly related to manifest expiration (which deletes the subs in candlepin but keeps the repo in Katello db)

This commit pro actively fails content view publish if any of the repositories attached to the content view does not have  a content object in candlepin
 
#### What are the testing steps for this pull request?
Not able to exactly reproduce this situation but here are some rails commands to try
1. Create a custom product and repository
2. `foreman-rake console`
3.
```ruby
> root = ::Katello::RootRepository.last
> ::Katello::Resources::Candlepin::Product.remove_content(root.product.organization.label, root.product.cp_id, root.content_id)
> ::Katello::Resources::Candlepin::Content.destroy(root.product.organization.label, root.content_id)
```
4. Add this repo to a content view and publish.

Before PR:
Error in finalize pausing the task and causing all kinds of side effects.

After PR:
Error asking the invalid repo to be removed before publish starts

### CODE REVIEW

File: app/models/katello/content_view.rb [hunk 1]
```diff
@@ -621,6 +622,32 @@ def check_docker_repository_names!(environments)
       true
     end
 
+    def check_repositories_exist!
+      bad_repo = repositories.yum_type.find do |repo|
+        begin
+          ::Katello::Resources::Candlepin::Content.get(organization.label, repo.root.content_id)
```
Comment (sjha4) [inline app/models/katello/content_view.rb:628]: Since this is a backend candlepin API call, would it make sense to move this to the publish task instead?

Comment (parthaa) [reply app/models/katello/content_view.rb]: Moved it to the CV helper

**Comment by sjha4** [comment]
Tested the composite part with missing content in a cv repo with 12 repos. Didn't notice much of a delay with the publish task starting and there's always the loading screen for users to see as they wait. 😸 

Screenshot from 2022-09-29 12-49-48


Ack from my end based on code and limited testing. Will let @chris1984 give the final ack/nack.

**Review by chris1984 — APPROVED** [review]
Works great, tested with 20 repos in a single content view since @sjha4 did the composite and it works great with the error and good wording. Lets ship it.

File: app/lib/actions/katello/repository/sync.rb [hunk 2]
```diff
@@ -25,11 +25,11 @@ def plan(repo, options = {})
           validate_contents = options.fetch(:validate_contents, false)
           skip_metadata_check = options.fetch(:skip_metadata_check, false) || (validate_contents && repo.yum?)
           generate_applicability =  options.fetch(:generate_applicability, repo.yum? || repo.deb?)
-
+          skip_candlepin_check = options.fetch(:skip_candlepin_check, false)
```
Comment (chris1984) [inline app/lib/actions/katello/repository/sync.rb:28]: This param is meant to be internal for us or support to debug right? If not should we expose that param on the controller side so hammer picks it up?

**Review by chris1984 — APPROVED** [review]
Tested with sync and promote, and looks amazing. 

I thought on the wording for a bit and can't really think of a way to improve it other than what we talked about yesterday, which you changed. One small question from me on the `skip_candlepin_check` but other than that it looks good.

File: app/lib/katello/util/candlepin_repository_checker.rb [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+  module Util
+    module CandlepinRepositoryChecker
+      def self.check_repositories_for_promote!(content_view_version)
+        bad_repo = content_view_version.repositories.yum_type.find { |repo| !repository_exist_in_backend?(repo) }
+        return if bad_repo.blank?
+        item = _("Content View Version: '%{cvv}', Product: '%{product}', Repository: '%{repo}' " %
+                  { repo: bad_repo.name, product: bad_repo.product.name, cvv: content_view_version.name })
+
+        if bad_repo.redhat?
+          fail _("%{item} does not have a valid subscription. "\
```
Comment (jeremylenz) [inline app/lib/katello/util/candlepin_repository_checker.rb:11]:
```suggestion
          fail _("'%{item}' does not exist in the backend system [ Candlepin ]. "\
```

File: app/lib/katello/util/candlepin_repository_checker.rb [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+                 " Either remove the invalid repository or try refreshing "\
+                 "the manifest before promoting. " % { item: item })
+        else
+          fail _("%{item} does not have a valid subscription. "\
```
Comment (jeremylenz) [inline app/lib/katello/util/candlepin_repository_checker.rb:15]:
```suggestion
          fail _("'%{item}' does not exist in the backend system [ Candlepin ]. "\
```

File: app/lib/katello/util/candlepin_repository_checker.rb [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+        end
+      end
+
+      def self.check_repository_for_sync!(repo)
+        return if !repo.yum? || repository_exist_in_backend?(repo)
+        item = _("Product: '%{product}', Repository: '%{repo}' " %
+                  { repo: repo.name, product: repo.product.name })
+
+        if repo.redhat?
+          fail _("%{item} does not have a valid subscription. "\
```
Comment (jeremylenz) [inline app/lib/katello/util/candlepin_repository_checker.rb:26]:
```suggestion
          fail _("'%{item}' does not exist in the backend system [ Candlepin ]. "\
```

File: app/lib/katello/util/candlepin_repository_checker.rb [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+                 " Either remove and re-enable the repository or try refreshing "\
+                 "the manifest before synchronizing. " % { item: item })
+        else
+          fail _("%{item} does not have a valid subscription. "\
```
Comment (jeremylenz) [inline app/lib/katello/util/candlepin_repository_checker.rb:30]:
```suggestion
          fail _("'%{item}' does not exist in the backend system [ Candlepin ]. "\
```

File: app/lib/katello/util/candlepin_repository_checker.rb [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+        return if bad_repo.blank?
+        if component_version
+          item = _("Component Version: '%{cvv}', Product: '%{product}', Repository: '%{repo}' " %
+                  { repo: bad_repo.name, product: bad_repo.product.name, cvv: component_version.name })
+        else
+          item = _("Product: '%{product}', Repository: '%{repo}' " %
+                  { repo: bad_repo.name, product: bad_repo.product.name })
+        end
+        if bad_repo.redhat?
+          fail _("%{item} in this content view does not have a valid subscription. "\
```
Comment (jeremylenz) [inline app/lib/katello/util/candlepin_repository_checker.rb:63]:
```suggestion
          fail _("'%{item}' in this content view does not exist in the backend system [ Candlepin ]. "\
```

File: app/lib/katello/util/candlepin_repository_checker.rb [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+                 " Either remove the invalid repository or try refreshing "\
+                 "the manifest before publishing again. " % { item: item })
+        else
+          fail _("%{item} in this content view does not have a valid subscription. "\
```
Comment (jeremylenz) [inline app/lib/katello/util/candlepin_repository_checker.rb:67]:
```suggestion
          fail _("'%{item}' in this content view does not exist in the backend system [ Candlepin ]. "\
```

**Review by jeremylenz — COMMENTED** [review]
Adding wording suggestion as requested (you may want to DRY it up a bit as well)

**Pull request merged**